iree/tools/utils/cpu_features.c

```diff
@@ -0,0 +1,173 @@
+#define IREE_CPU_FEATURES_AARCH64_READ_REGISTER(REG)               \
+  static uint64_t iree_cpu_features_aarch64_read_##REG(void) {     \
+    int64_t retval = 0;                                            \
+    asm("mrs %[dst], " #REG                                        \
+      : /* outputs */                                              \
+      [dst] "=r"(retval)                                           \
+      : /* inputs */                                               \
+      : /* clobbers */);                                           \
+    return retval;                                                 \
+  }
+// clang-format on
+
+// Define the access functions for the cpu feature registers that we need.
+IREE_CPU_FEATURES_AARCH64_READ_REGISTER(ID_AA64ISAR0_EL1)
+IREE_CPU_FEATURES_AARCH64_READ_REGISTER(ID_AA64ISAR1_EL1)
+
+// Helper for iree_cpu_features_aarch64_get_register:
+// Immediately reads the value of a cpu feature register, by enum index.
+// Expensive!
+static uint64_t iree_cpu_features_aarch64_read_register(
+    iree_cpu_features_aarch64_register_t idx) {
+  switch (idx) {
+    case iree_cpu_features_aarch64_register_ID_AA64ISAR0_EL1:
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR0_EL1();
+    case iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1:
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR1_EL1();
+    case iree_cpu_features_aarch64_register_count:  // case, not default, so
+                                                    // that
+      break;  // the compiler can warn if we forget an enum value.
+  }
+  assert(false && "bad iree_cpu_features_aarch64_register_t value");
+  return 0;
+}
+
+// Returns the value of a cpu feature register. Caches the value in the
+// iree_cpu_features_t struct.
+static uint64_t iree_cpu_features_aarch64_get_register(
+    iree_cpu_features_t* features, iree_cpu_features_aarch64_register_t idx) {
+  if (!features->registers_initialized[idx]) {
+    if (iree_devices_features_aarch64_register_access()) {
+      features->registers[idx] = iree_cpu_features_aarch64_read_register(idx);
+    } /* else: leave registers[idx] with its initial 0 bits */
+    features->registers_initialized[idx] = true;
+  }
+  return features->registers[idx];
+}
+
+// Returns true if +dotprod is supported.
+static bool iree_cpu_features_aarch64_dotprod(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR0_EL1);
+  return (register_value >> 44) & 1;
+}
+
+// Returns true if +i8mm is supported.
+static bool iree_cpu_features_aarch64_i8mm(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1);
+  return (register_value >> 52) & 1;
+}
+
+typedef iree_cpu_features_t iree_cpu_features_t;
+
+#else  // other CPU architectures
+
+struct iree_cpu_features_t {};
```

Comment:
quick nit but can review in more detail next week: empty structs are not well supported on all compilers. We usually add a 'reserved' field in those cases.

Reply:
oh ok! i'm fine with adding an unused field to make it non-empty, and it will allow to substantially simplify both the logic (all those `if sizeof == 0` checks) and the comments (had to comment the peculiar semantics of the zero-sized case, returning success but setting the pointer to null --- now i wont need any special-casing anymore).

the downside is of course that the vacuous non-implemented case will now do an allocator allocation, which we were avoiding for now. negligible overhead, but i guess it was neat to avoid all allocator work, e.g. it was avoiding polluting any allocator logging/tracing that one might want to do, with unusued tiny allocations.

Reply:
if we care about that latter aspect, we can still recover that with a non-empty struct, by introducing a `#define IREE_CPU_FEATURES_IMPLEMENTED` to check for instead of checking the struct size.

@benvanik @ScottTodd  let me know what you prefer.

Reply:
OK, i've just added `int unused` and stopped worrying about tiny allocs. The diff is satisfyingly red, thanks for the suggestion.